### PR TITLE
feat: dp and pt: implement exclude types in descriptor se_a

### PR DESCRIPTION
**Pull request by wanghan-iapcm**

### CODE REVIEW

File: source/tests/pt/model/test_se_e2_a.py [hunk 1]
```diff
@@ -40,9 +40,10 @@ def test_consistency(
         dstd = rng.normal(size=(self.nt, nnei, 4))
         dstd = 0.1 + np.abs(dstd)
 
-        for idt, prec in itertools.product(
+        for idt, prec, em in itertools.product(
             [False, True],
             ["float64", "float32"],
+            [[[0, 1]], [[1, 1]]],
```
Comment (njzjz) [inline source/tests/pt/model/test_se_e2_a.py:46]: Should we keep the test for `[]` (the original case)?

Comment (wanghan-iapcm) [reply source/tests/pt/model/test_se_e2_a.py:46]: sure

**Comment by njzjz** [comment]
GPU tests failed.

File: deepmd/pt/model/descriptor/descriptor.py [hunk 2]
```diff
@@ -206,6 +208,32 @@ class DescriptorBlock(torch.nn.Module, ABC):
...
+                [
+                    1 if (tt_i, tt_j) not in _exclude_types else 0
+                    for tt_i in range(ntypes + 1)
+                ]
+                for tt_j in range(ntypes + 1)
+            ],
+            dtype=np.int32,
+        )
+        # (ntypes+1 x ntypes+1)
+        self.type_mask = torch.from_numpy(self.type_mask).view([-1])
```
Comment (iProzd) [inline deepmd/pt/model/descriptor/descriptor.py:234]: Should `self.type_mask` be saved through `register_buffer`, so it can be saved into `model.pt`? Or we allow user to change `exclude_types` each time when restart from saved checkpoint, even if it doesn't match each other?

Comment (iProzd) [reply deepmd/pt/model/descriptor/descriptor.py:234]: Additionally, `device` missed here for `torch.from_numpy`, GPU may fail.

File: deepmd/pt/model/descriptor/descriptor.py [hunk 3]
```diff
@@ -332,7 +360,50 @@ def forward(
...
         """Calculate DescriptorBlock."""
-        raise NotImplementedError
+        pass
+
+    # may have a better place for this method...
+    def build_type_exclude_mask(
+        self,
+        nlist: torch.Tensor,
+        atype_ext: torch.Tensor,
+    ) -> torch.Tensor:
```
Comment (iProzd) [inline deepmd/pt/model/descriptor/descriptor.py:370]: Maybe we can make `ExcludeMask` class in utils and as an attribute (if needed) in atomic model level? Then we can get `exlude_mask` at the begining of model `forward` and pass the mask to `forward` of both descriptor and fitting.

Comment (iProzd) [reply deepmd/pt/model/descriptor/descriptor.py:370]: Anyway, this can also be done in spin PR.

Comment (wanghan-iapcm) [reply deepmd/pt/model/descriptor/descriptor.py:373]: A `ExcludeMask` class like that in the `dpmodel` cannot be jitted...
the exclude_mask is only needed by descriptors, because it is pairwise. The fitting needs only single atom masks.

**Review by njzjz — APPROVED** [review]

**Review by iProzd — APPROVED** [review]
